src/ucs/datastruct/hlist.h

```diff
@@ -165,6 +165,57 @@ ucs_hlist_extract_head(ucs_hlist_head_t *head)
 }
 
 
+/**
+ * Get list head element as the containing type, assuming the list is not empty.
+ *
+ * @param _head    List head.
+ * @param _type    Containing structure type.
+ * @param _member  List element inside the containing structure.
+ *
+ * @note If the list is empty this macro has undefined behavior.
+ */
+#define ucs_hlist_head_elem(_head, _type, _member) \
+    ucs_container_of((_head)->ptr, _type, _member)
+
+
+/**
+ * Get list next element as the containing type.
+ *
+ * @param _elem    List element.
+ * @param _type    Containing structure type.
+ * @param _member  List element inside the containing structure.
+ */
+#define ucs_hlist_next_elem(_elem, _member) \
+    ucs_container_of(ucs_list_next(&(_elem)->_member.list, ucs_hlist_link_t, \
+                                   list), \
+                     typeof(*(_elem)), _member)
+
+
+/**
+ * Iterate over detached-head list.
+ *
+ * @param _elem     Variable to hold the current list element
+ * @param _head     Pointer to list head.
+ * @param _member   List element inside the containing structure.
+ *
+ * @note The iteration is implemented by first setting the element to NULL, then
+ * inside 'for' loop condition (which is done before each iteration), we advance
+ * the element pointer and check for end condition: in the first iteration,
+ * when elem is NULL, we check that the list is not empty. For subsequent
+ * iterations, we check that elem has not reached the list head yet.
+ */
+#define ucs_hlist_for_each(_elem, _head, _member) \
+    for (_elem = NULL; \
```

Comment:
can use the approach from https://github.com/openucx/ucx/commit/44a8b8a77dd6af49b944a1275249d53f00d70239 to address [CPPCHECK warning](https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=5817&view=logs&j=c3b1c751-b3c6-5682-a32e-d1fa8fde7511&t=2a70c147-fc75-51ad-876a-fada1a13068c):
```
for (_elem = NULL; \
     ...
      ((_elem = ucs_hlist_head_elem(_head, typeof(*(_elem)), _member)) != \
                 ucs_container_of(_head, typeof(*(_elem)), _member)) : \
```
and set `_head->ptr` to `_head` - to fix all places where using `ucs_hlist_head_elem()` for `_head->ptr == NULL`
wdyt?

Reply:
@dmitrygx thanks for the idea, but unfortunately, if we use `_head` as replacement for NULL, it makes hlist not possible to add to a resizing hash like khash (one of the goals of hlist is to make it possible to add to hash)

Reply:
@yosefe I see